### PR TITLE
drivers: serial: lis3mdl: support for SPI

### PR DESCRIPTION
**Pull request by DeMurlot**
This pull request add two things to the lis3mdl driver
- support for usage through SPI
- Kconfig option to enable die temperature reading

The first part is heavily inspired by the lsm6dsl's way of supporting both I2C and SPI.

### CODE REVIEW

**Comment by DeMurlot** [comment]
The current Kconfig default is to disable support for die temperature measurement and breaks backwards comparability. Should I enable it by default?

**Comment by jeppenodgaard** [comment]
> The current Kconfig default is to disable support for die temperature measurement and breaks backwards comparability. Should I enable it by default?

I think it is safest to keep it as is.

If it is changed I think that change should have its own commit, and maybe probably be mentioned in the 4.0 migration guide.

**Comment by jeppenodgaard** [comment]
The commits needs a body as described in:
https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines

File: drivers/sensor/st/lis3mdl/lis3mdl.c [hunk 1]
```diff
@@ -85,11 +91,12 @@ int lis3mdl_sample_fetch(const struct device *dev, enum sensor_channel chan)
 		LOG_DBG("Failed to fetch temperature sample.");
 		return -EIO;
 	}
+	drv_data->temp_sample = sys_le16_to_cpu(buf[3]);
```
Comment (avisconti) [inline drivers/sensor/st/lis3mdl/lis3mdl.c:93]: where did you read these two bytes?

**Review by avisconti — CHANGES_REQUESTED** [review]
The first commit looks ok a part a couple of comments

File: dts/bindings/sensor/st,lis3mdl-spi.yaml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2017, Linaro Limited
```
Comment (avisconti) [inline dts/bindings/sensor/st,lis3mdl-spi.yaml:1]: 2024 and maybe not Linaro

File: dts/bindings/sensor/st,lis3mdl-i2c.yaml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2017, Linaro Limited
```
Comment (avisconti) [inline dts/bindings/sensor/st,lis3mdl-i2c.yaml:1]: 2024 and Linaro?

File: drivers/sensor/st/lis3mdl/lis3mdl.c [hunk 4]
```diff
@@ -172,17 +174,73 @@ int lis3mdl_init(const struct device *dev)
...
+		.bus_cfg.i2c = I2C_DT_SPEC_INST_GET(inst),		\
+		COND_CODE_1(DT_INST_NODE_HAS_PROP(inst, irq_gpios),	\
+		(LIS3MDL_CFG_IRQ(inst)), ())				\
+	}
+
+#define LIS3MDL_DEFINE_I2C(inst)					\
+	static struct lis3mdl_data lis3mdl_data_##inst;			\
+	static struct lis3mdl_config lis3mdl_config_##inst =	\
+		LIS3MDL_CONFIG_I2C(inst);				\
+	LIS3MDL_DEVICE_INIT(inst)
```
Comment (avisconti) [inline drivers/sensor/st/lis3mdl/lis3mdl.c:239]: this macro is common between i2c and spi, so maybe you can call only once in ```LIS3MDL_DEFINE()``` as some other st driver do.

File: drivers/sensor/st/lis3mdl/lis3mdl.c [hunk 5]
```diff
@@ -172,17 +174,73 @@ int lis3mdl_init(const struct device *dev)
...
+/*
+ * Instantiation macros used when a device is on an I2C bus.
+ */
+
+#define LIS3MDL_CONFIG_I2C(inst)					\
+	{								\
+		.bus_init = lis3mdl_i2c_init,				\
+		.bus_cfg.i2c = I2C_DT_SPEC_INST_GET(inst),		\
+		COND_CODE_1(DT_INST_NODE_HAS_PROP(inst, irq_gpios),	\
+		(LIS3MDL_CFG_IRQ(inst)), ())				\
```
Comment (avisconti) [inline drivers/sensor/st/lis3mdl/lis3mdl.c:232]: same for i2c and spi. Maybe call only once?

**Review by avisconti — CHANGES_REQUESTED** [review]
maybe we could move to stmemsc directly instead of introducing hw_tf?

**Comment by avisconti** [comment]
@DeMurlot I guess you may also want to add the spi test for the sensor in ```tests/drivers/build_all/sensor/```

File: dts/bindings/sensor/st,lis3mdl-spi.yaml (hunk 2 above)
Comment (DeMurlot) [reply dts/bindings/sensor/st,lis3mdl-spi.yaml]: fixed, thank you

File: dts/bindings/sensor/st,lis3mdl-i2c.yaml (hunk 3 above)
Comment (DeMurlot) [reply dts/bindings/sensor/st,lis3mdl-i2c.yaml]: fixed

**Pull request closed without merging**